sdk/cosmosdb/cosmos/src/globalPartitionEndpointManager.ts

```diff
@@ -0,0 +1,529 @@
+      // Un-deterministically marking the original failed endpoint for the PkRange back to healthy.
+      pkRangeToEndpointMappings.set(pkRange, [originalFailedLocation, HealthStatus.Connected]);
+    }
+    return;
+  }
+}
+/**
+ * @hidden
+ */
+export class PartitionKeyRangeFailoverInfo {
+  private failedEndPoints: string[] = [];
+  public currentEndPoint: string;
+  public firstFailedEndPoint: string;
+
+  private consecutiveReadRequestFailureCount: number = 0;
+  private consecutiveWriteRequestFailureCount: number = 0;
+  private firstRequestFailureTime: number = Date.now();
+  private lastRequestFailureTime: number = Date.now();
+
+  private failureCountSemaphore: semaphore.Semaphore;
+  private timestampSemaphore: semaphore.Semaphore;
+  private tryMoveNextLocationSemaphore: semaphore.Semaphore;
+
+  /**
+   * @internal
+   */
+  constructor(currentEndpoint: string) {
+    this.currentEndPoint = currentEndpoint;
+    this.firstFailedEndPoint = currentEndpoint;
+    this.failureCountSemaphore = semaphore(1);
+    this.timestampSemaphore = semaphore(1);
+    this.tryMoveNextLocationSemaphore = semaphore(1);
+  }
+
+  public async CanCircuitBreakerTriggerPartitionFailOver(
+    isReadOnlyRequest: boolean,
+  ): Promise<boolean> {
+    const { consecutiveReadRequestFailureCount, consecutiveWriteRequestFailureCount } =
+      await this.snapshotConsecutiveRequestFailureCount();
+
+    return isReadOnlyRequest
+      ? consecutiveReadRequestFailureCount > Constants.ReadRequestFailureCounterThreshold
+      : consecutiveWriteRequestFailureCount > Constants.WriteRequestFailureCounterThreshold;
+  }
+
+  public async incrementRequestFailureCounts(
+    isReadOnlyRequest: boolean,
+    currentTime: number,
+  ): Promise<void> {
+    return new Promise<void>((resolve, reject) => {
+      this.failureCountSemaphore.take(async () => {
+        try {
+          const { lastRequestFailureTime } = await this.snapshotPartitionFailoverTimestamps();
+
+          if (currentTime - lastRequestFailureTime > Constants.TimeoutCounterResetWindow) {
+            this.consecutiveReadRequestFailureCount = 0;
+            this.consecutiveWriteRequestFailureCount = 0;
+          }
+
+          if (isReadOnlyRequest) {
+            this.consecutiveReadRequestFailureCount++;
+          } else {
+            this.consecutiveWriteRequestFailureCount++;
+          }
+          this.lastRequestFailureTime = currentTime;
+          return resolve();
+        } catch (error) {
+          reject(error);
+        } finally {
+          // Release the semaphore lock
+          this.failureCountSemaphore.leave();
+        }
+      });
+    });
+  }
+
+  public async snapshotConsecutiveRequestFailureCount(): Promise<{
+    consecutiveReadRequestFailureCount: number;
+    consecutiveWriteRequestFailureCount: number;
+  }> {
+    return new Promise((resolve, reject) => {
+      this.failureCountSemaphore.take(() => {
```

Comment:
this is just a read call, so what is the point of using a semaphore here? Why do we want to restrict reads using semaphore.